### PR TITLE
Update evaluate_postfix_notations.py 

### LINKED ISSUE #8754: Need to support for unary operator in postfix evaluation.
### Feature description

As mentioned in #8724, we have two files for evaluating postfix expressions. But currently both of them only work with binary operators and will raise an exception when an unary operator is given.

example:
```python
>>> res = evaluate_postfix(["2", "-", "3", "+"])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../stacks/evaluate_postfix_notations.py", line 31, in evaluate_postfix
    b, a = stack.pop(), stack.pop()
IndexError: pop from empty list
>>> 
```
`2-3+` is a postfix notation for `-2+3` infix and should result in `1`.

### PR DESCRIPTION
**Pull request by rohan472000**
### Describe your change:

Fixes #8754

* [ ] Add an algorithm?
* [x] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: {#8754}`.

### CODE REVIEW

**Comment by amirsoroush** [comment]
Hi, Please add `Fixes #8754` to the PR's description so that GitHub can attach this PR to the issue.

File: data_structures/stacks/evaluate_postfix_notations.py [hunk 1]
```diff
@@ -17,32 +17,52 @@ def evaluate_postfix(postfix_notation: list) -> int:
...
+    >>> evaluate_postfix(["5", "-"])
+    -5
     >>> evaluate_postfix([])
     0
     """
     if not postfix_notation:
         return 0
 
     operations = {"+", "-", "*", "/"}
+    unary_operations = {"+"}  # Unary operator(s)
```
Comment (amirsoroush) [inline data_structures/stacks/evaluate_postfix_notations.py:31]: `unary_operations` is not currently being used.

**Review by amirsoroush — COMMENTED** [review]
I couldn't pass the doctests. Here are two failed test cases:

```none
**********************************************************************
File ".../evaluate_postfix_notations.py", line 18, in __main__.evaluate_postfix
Failed example:
    evaluate_postfix(["4", "13", "5", "/", "+"])
Expected:
    6
Got:
    6.6
**********************************************************************
File ".../evaluate_postfix_notations.py", line 20, in __main__.evaluate_postfix
Failed example:
    evaluate_postfix(["2", "+"])
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib/python3.10/doctest.py", line 1350, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest __main__.evaluate_postfix[2]>", line 1, in <module>
        evaluate_postfix(["2", "+"])
      File ".../evaluate_postfix_notations.py", line 46, in evaluate_postfix
        raise ValueError(
    ValueError: Invalid expression: insufficient operands for binary operator
**********************************************************************
1 items had failures:
   2 of   5 in __main__.evaluate_postfix
***Test Failed*** 2 failures.
```

File: data_structures/stacks/evaluate_postfix_notations.py [hunk 2]
```diff
@@ -17,32 +17,52 @@ def evaluate_postfix(postfix_notation: list) -> int:
...
-                stack.append(a + b)
-            elif token == "-":
-                stack.append(a - b)
-            elif token == "*":
-                stack.append(a * b)
+            if token == "-":
+                if len(stack) < 1:
+                    raise ValueError(
+                        "Invalid expression: insufficient operands for unary operator"
+                    )
```
Comment (tianyizheng02) [inline data_structures/stacks/evaluate_postfix_notations.py:41]: This doesn't fix the problem with unary negation operators. We want to support negation, not raise an error if it's present.

**Review by tianyizheng02 — CHANGES_REQUESTED** [review]

File: data_structures/stacks/evaluate_postfix_notations.py (hunk 1 above)
Comment (rohan472000) [reply data_structures/stacks/evaluate_postfix_notations.py]: now I made several changes, kindly look into it.

File: data_structures/stacks/evaluate_postfix_notations.py (hunk 2 above)
Comment (rohan472000) [reply data_structures/stacks/evaluate_postfix_notations.py]: made several changes, look into that.

File: data_structures/stacks/evaluate_postfix_notations.py [hunk 3]
```diff
@@ -28,17 +32,23 @@ def evaluate_postfix(postfix_notation: list) -> int:
...
-            elif token == "-":
-                stack.append(a - b)
-            elif token == "*":
-                stack.append(a * b)
+            if token == "-" and len(stack) < 2:
+                operand = stack.pop()
+                stack.append(-operand)
+            elif (True) and len(stack) < 2:
+                operand = stack.pop()
+                stack.append(operand)
```
Comment (tianyizheng02) [inline data_structures/stacks/evaluate_postfix_notations.py:40]: What is the purpose of this `elif` block?

Comment (rohan472000) [reply data_structures/stacks/evaluate_postfix_notations.py:40]: This block will handle +,*,/ operators as above block is handling "-" operator.

**Review by amirsoroush — CHANGES_REQUESTED** [review]
I guess that way is more readable.

**Comment by rohan472000** [comment]
@cclauss, look into this issue in order to merge.

**Review by cclauss — CHANGES_REQUESTED** [review]
Doing this work without using https://docs.python.org/3/library/operator.html seems suboptimal.

We should be able to deal with both floats and ints.

Please add tests for negative numbers, zeros, floats and
* [ ] `evaluate_postfix(["5"])`
* [ ] `evaluate_postfix(["-"])`
* [ ] `evaluate_postfix(["5", "A"])`
* [ ] `evaluate_postfix(["A"])`

File: data_structures/stacks/evaluate_postfix_notations.py [hunk 4]
```diff
@@ -28,17 +32,22 @@ def evaluate_postfix(postfix_notation: list) -> int:
...
-            if token == "+":
-                stack.append(a + b)
-            elif token == "-":
-                stack.append(a - b)
-            elif token == "*":
-                stack.append(a * b)
+            if len(stack) < 2:
+                operand = stack.pop()
+                if token == "-":
+                    operand = -operand
```
Comment (cclauss) [inline data_structures/stacks/evaluate_postfix_notations.py:38]:
```suggestion
                    operand = -operand
                else:
                    raise ValueError(f"Unrecognized {token = }")
```

File: data_structures/stacks/evaluate_postfix_notations.py [hunk 5]
```diff
@@ -28,17 +32,22 @@ def evaluate_postfix(postfix_notation: list) -> int:
...
+                if token == "+":
+                    stack.append(a + b)
+                elif token == "-":
+                    stack.append(a - b)
+                elif token == "*":
+                    stack.append(a * b)
+                elif token == "/":
+                    if b == 0:
+                        raise ValueError("Invalid expression: division by zero")
                     stack.append(a // b)
```
Comment (cclauss) [inline data_structures/stacks/evaluate_postfix_notations.py:51]:
```suggestion
                    stack.append(a // b)
                else:
                    raise ValueError(f"Unrecognized {token = }")
```

File: data_structures/stacks/evaluate_postfix_notations.py [hunk 6]
```diff
@@ -17,9 +17,29 @@ def evaluate_postfix(postfix_notation: list) -> int:
...
+    Traceback (most recent call last):
+     ...
+    ValueError: invalid literal for int() with base 10: 'A'
+    >>> evaluate_postfix(["A"])
+    Traceback (most recent call last):
+     ...
+    ValueError: invalid literal for int() with base 10: 'A'"""
+    if len(postfix_notation) == 1 and postfix_notation[0] in {"+", "-", "*", "/"}:
+        raise ValueError("Invalid expression: insufficient operands")
+
```
Comment (tianyizheng02) [inline data_structures/stacks/evaluate_postfix_notations.py:42]:
```suggestion
```
This can be handled by simplifying logic elsewhere. Plus, listing out the set of operators is redundant with later code.

File: data_structures/stacks/evaluate_postfix_notations.py [hunk 7]
```diff
@@ -28,18 +48,27 @@ def evaluate_postfix(postfix_notation: list) -> int:
...
-                stack.append(a * b)
-            else:
-                if a * b < 0 and a % b != 0:
-                    stack.append(a // b + 1)
+            if len(stack) < 2:
+                operand = stack.pop()
+                if token == "-":
+                    stack.append(-operand)
                 else:
+                    stack.append(operand)
```
Comment (tianyizheng02) [inline data_structures/stacks/evaluate_postfix_notations.py:56]:
```suggestion
            if len(stack) == 0:
                msg = f"{token} operator with no operand(s)"
                raise ValueError(msg)
            if token == "-" and len(stack) < 2:
                operand = stack.pop()
                stack.append(-operand)
```

**Review by tianyizheng02 — CHANGES_REQUESTED** [review]

File: data_structures/stacks/evaluate_postfix_notations.py [hunk 8]
```diff
@@ -28,18 +32,28 @@ def evaluate_postfix(postfix_notation: list) -> int:
...
-            else:
-                if a * b < 0 and a % b != 0:
-                    stack.append(a // b + 1)
+            if len(stack) < 2:
+                operand = stack.pop()
+                if token == "-":
+                    operand = -operand
                 else:
+                    raise ValueError(f"Unrecognized {token = }")
+                stack.append(operand)
```
[the remark this answers is not in the document]
Comment (rohan472000) [reply data_structures/stacks/evaluate_postfix_notations.py]: I'm not able to commit above change requested...!!

**Comment by rohan472000** [comment]
@tianyizheng02 ,as your suggestion is making build fail(doctest not passing), so I have again pushed my previous code.

**Comment by tianyizheng02** [comment]
The build was failing because this test was failing:
```py
>>> evaluate_postfix(["2", "+"])
```
```
UNEXPECTED EXCEPTION: IndexError('pop from empty list')
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.11.3/x64/lib/python3.11/doctest.py", line 1351, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest data_structures.stacks.evaluate_postfix_notations.evaluate_postfix[2]>", line 1, in <module>
  File "/home/runner/work/Python/Python/data_structures/stacks/evaluate_postfix_notations.py", line 55, in evaluate_postfix
    b, a = stack.pop(), stack.pop()
                        ^^^^^^^^^^^
IndexError: pop from empty list
/home/runner/work/Python/Python/data_structures/stacks/evaluate_postfix_notations.py:20: UnexpectedException
```
I believe my suggestion was still correct, but you need to add an if-statement to handle this edge case as well. This test should raise an exception because `+` is a binary operator and there is only 1 operand.

**Comment by rohan472000** [comment]
ok...but main concern is that for `>>> evaluate_postfix(["-"])` case....which is still not resolved..

**Comment by tianyizheng02** [comment]
> ok...but main concern is for `>>> evaluate_postfix(["-"])` case....which is still not resolved..

I believe that test was already passing. Again, my suggestion (specifically the `elif len(stack) == 0:` line) should've handled it.

**Comment by rohan472000** [comment]
> > ok...but main concern is for `>>> evaluate_postfix(["-"])` case....which is still not resolved..
> 
> I believe that test should've already been passing. Again, my suggestion (specifically the `elif len(stack) == 0:` line) should've handled it.

**Comment by tianyizheng02** [comment]
Issue has already been fixed by another PR
